src/Codec/Xlsx/Writer/Stream.hs

```diff
@@ -284,47 +293,44 @@ setNameSpaceRec space xelm =
                                     y -> y
     }
 
-columns :: Monad m => WriteSettings -> ConduitT SheetItem Event m ()
-columns settings =
+columns :: MonadReader WriteSettings m => ConduitT SheetItem Event m ()
+columns = do
+  colProps <- view wsColumnProperties
+  let cols :: Maybe TXML.Element
+      cols = nonEmptyElListSimple (n_ "cols") $ map (toElement (n_ "col")) colProps
   traverse_ (C.yieldMany . elementToEvents . toXMLElement) cols
-  where
-    cols :: Maybe TXML.Element
-    cols = nonEmptyElListSimple (n_ "cols") . map (toElement (n_ "col")) $ settings ^. wsColumnProperties
-
-writeWorkSheet :: Monad m => WriteSettings -> Map Text Int  -> ConduitT SheetItem Event m ()
-writeWorkSheet settings sstable = doc (n_ "worksheet") $ do
-    sheetViews settings
-    columns settings
-    el (n_ "sheetData") $ C.concatMap (mapItem sstable)
-
-
-mapItem :: Map Text Int -> SheetItem -> [Event]
-mapItem sstable sheetItem =
-  [EventBeginElement (n_ "row")  [("r", [ContentText $ toAttrVal rowIx])]]
-   <>
-  (ifoldMap (mapCell sstable rowIx) $ sheetItem ^. si_cell_row)
-   <>
-  [EventEndElement (n_ "row")]
+
+writeWorkSheet :: MonadReader WriteSettings  m => Map Text Int  -> ConduitT SheetItem Event m ()
+writeWorkSheet sstable = doc (n_ "worksheet") $ do
+    sheetViews
+    columns
+    el (n_ "sheetData") $ C.awaitForever (mapRow sstable)
+
+mapRow :: MonadReader WriteSettings m => Map Text Int -> SheetItem -> ConduitT SheetItem Event m ()
+mapRow sstable sheetItem = do
+  mRowProp <- preview $ wsRowProperties . ix rowIx . rowHeightLens . _Just . failing _CustomHeight _AutomaticHeight
+  let rowAttr :: Attributes
+      rowAttr = ixAttr <> fold (attr "ht" . txtd <$> mRowProp)
+  tag (n_ "row") rowAttr $
+    void $ itraverse (mapCell sstable rowIx) (sheetItem ^. si_cell_row)
   where
     rowIx = sheetItem ^. si_row_index
+    ixAttr = attr "r" $ toAttrVal rowIx
 
-mapCell :: Map Text Int -> RowIndex -> ColIndex -> Cell -> [Event]
+mapCell :: Monad m => Map Text Int -> RowIndex -> ColIndex -> Cell -> ConduitT SheetItem Event m ()
```

Comment:
The second part of this refactor changes this Event list into a conduit, which removes quite a few `<>`, which can be slow for list. Furthermore the implementation becomes a lot shorter because now we can use the render functions `tag` and `el`.